packages/editor/src/components/post-publish-panel/maybe-tags-panel.js

```diff
@@ -0,0 +1,74 @@
+/**
+ * WordPress dependencies
+ */
+import { __ } from '@wordpress/i18n';
+import { Component } from '@wordpress/element';
+import { compose } from '@wordpress/compose';
+import { withSelect } from '@wordpress/data';
+import { Dashicon, PanelBody } from '@wordpress/components';
+
+import FlatTermSelector from '../post-taxonomies/flat-term-selector';
+
+const TagsPanel = () => <PanelBody initialOpen={ true } title={ [
+	<Dashicon key={ 'dashicon-lightbulb' } icon={ 'lightbulb' } />,
+	__( 'Tip:' ),
+	<span className="editor-post-publish-panel__link" key="label">{
+		__( 'Add tags to your post' )
+	}</span>,
+] }>
+	<FlatTermSelector slug={ 'post_tag' } />
+</PanelBody>;
+
+class MaybeTagsPanel extends Component {
+	constructor( props ) {
+		super( props );
+		this.state = {
+			hadTags: props.hasTags,
+		};
+	}
+
+	/*
+	 * We use state.hadTags to track whether the post had tags when this
+	 * component was first mounted.
+	 *
+	 * We can't rely on the props.hasTags value to do this check
+	 * because when the user adds a tag to a post that didn't have one,
+	 * the props.hasTags is going to change from false to true,
+	 * forcing a re-render of this component and hiding it when it was previosly shown.
+	 *
+	 * We take advantage of the fact that the props.hasTags value is undefined
+	 * if tags weren't fetched yet, otherwise it will be true of false.
+	 *
+	 */
+	static getDerivedStateFromProps( props, state ) {
+		const { hadTags: currentHadTags } = state;
+		const { hasTags } = props;
+		if ( currentHadTags === undefined && hasTags !== undefined ) {
+			return {
+				hadTags: hasTags,
+			};
+		}
+		return null;
+	}
+
+	render() {
+		const { isPostTypeSupported } = this.props;
+		const { hadTags } = this.state;
+		if ( ! isPostTypeSupported || hadTags ) {
+			return null;
+		}
+		return ( <TagsPanel /> );
+	}
+}
+
+export default compose( [
+	withSelect( ( select ) => {
+		const postType = select( 'core/editor' ).getCurrentPostType();
+		const tags = select( 'core' ).getTaxonomy( 'post_tag' );
+		const terms = tags ? select( 'core/editor' ).getEditedPostAttribute( tags.rest_base ) : [];
+		return {
+			hasTags: tags && terms && terms.length > 0,
```

Comment:
Can't this condition be something like that instead:

```
const tagTaxonomy = select( 'core' ).getTaxonomy( 'post_tag' );
const shouldShowMoreTags = tagTaxonomy && ( ! select( 'core/editor' ).getEditedPostAttribute( tags.rest_base ) || select( 'core/editor' ).getEditedPostAttribute( tags.rest_base ).length ===0 ) &&  tagTaxonomy.types.some( ( type ) => type === postType );
```

Something in that vein?

And I guess we can leverage the `ifCondition` HigherOrderComponent to avoid the `MaybeTagsPanel` component entirely?

Reply:
Thanks for the `ifCondition` suggestion: I've used that in e66d09d to prevent the component from mounting if some conditions aren't met.

In this case, the internal component state is useful: we need a place to encode whether the post had tags *when the user clicked publish*. We can't use the `hasTags` prop because it'll become `true` if the user adds a new tag in the pre-publish panel, causing a re-render that hides the tags component, and so preventing the user from adding more than one tag.

I've also refactored the logic to make the assumptions more explicit.

Reply:
Oh ok, I get it now and in that case, I think you were right originally and now it can fail if the taxonomy is not loaded initially (in theory, because in practice, it's already there).

But I'm wondering if the `hasTags` condition makes sense at all, why can't we just always show the tags there if the post supports it? It seems better for me even if I already entered tags to check them on pre-publish. It also makes the code way simpler.

Reply:
> Oh ok, I get it now and in that case, I think you were right originally and now it can fail if the taxonomy is not loaded initially (in theory, because in practice, it's already there).

It's still correct. Note that if the taxonomy is not loaded (which do happen when the user clicks the publish button without the sidebar being opened first), the `areTagsFetched` will be false. https://github.com/WordPress/gutenberg/commit/e66d09dc547fed395e7e9182250f9c451aaf33f4 decoupled two separate concerns that were previously encoded in only one variable: whether or not the tags taxonomy has been fetched (`areTagsFetched`) and if the post has tags (`hasTags`).

> It seems better for me even if I already entered tags to check them on pre-publish. It also makes the code way simpler.

That's a design decision that I defer to Matías and Sarah. It seems sensible to limit the amount of info in the pre-publish panel - we don't want to have here info that's already filled in the sidebar by the user.